File: app/src/main/java/com/carlos/grabredenvelope/receiver/JpushReceiver.java
package com.carlos.grabredenvelope.receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;

import com.carlos.grabredenvelope.main.MainActivity;
import com.carlos.grabredenvelope.util.PreferencesUtils;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Iterator;

import cn.jpush.android.api.JPushInterface;

/**
 * Created by 小不点 on 2016/2/22.
 */
/**
 * 自定义接收器
 *
 * 如果不定义这个 Receiver，则：
 * 1) 默认用户会打开主界面
 * 2) 接收不到自定义消息
 */
public class JpushReceiver extends BroadcastReceiver {
    private static final String TAG = "JpushReceiver";

    @Override
    public void onReceive(Context context, Intent intent) {
        Bundle bundle = intent.getExtras();
        Log.d(TAG, "[MyReceiver] onReceive - " + intent.getAction() + ", extras: " + printBundle(bundle));

        if (JPushInterface.ACTION_REGISTRATION_ID.equals(intent.getAction())) {
            String regId = bundle.getString(JPushInterface.EXTRA_REGISTRATION_ID);
            Log.d(TAG, "[MyReceiver] 接收Registration Id : " + regId);
            //send the Registration Id to your server...

        } else if (JPushInterface.ACTION_MESSAGE_RECEIVED.equals(intent.getAction())) {
            Log.d(TAG, "[MyReceiver] 接收到推送下来的自定义消息: " + bundle.getString(JPushInterface.EXTRA_MESSAGE));
//            processCustomMessage(context, bundle);
            String message=bundle.getString(JPushInterface.EXTRA_MESSAGE);
            if(message.equals("start")){
                Log.d(TAG,"可以用");
                PreferencesUtils.setUseStatus(true);
            }else if(message.equals("stop")){
                Log.d(TAG,"不可以使用");
                PreferencesUtils.setUseStatus(false);
            }

        } else if (JPushInterface.ACTION_NOTIFICATION_RECEIVED.equals(intent.getAction())) {
            Log.d(TAG, "[MyReceiver] 接收到推送下来的通知");
            int notifactionId = bundle.getInt(JPushInterface.EXTRA_NOTIFICATION_ID);
            Log.d(TAG, "[MyReceiver] 接收到推送下来的通知的ID: " + notifactionId);

        } else if (JPushInterface.ACTION_NOTIFICATION_OPENED.equals(intent.getAction())) {
            Log.d(TAG, "[MyReceiver] 用户点击打开了通知");
            String json=intent.getExtras().getString(JPushInterface.EXTRA_EXTRA);
            Log.d(TAG,"[object]"+json);
            try {
                JSONObject object=new JSONObject(json);
                Iterator<String> it=object.keys();
                while(it.hasNext()){
                    String key=it.next();
                    Log.d(TAG,"[object]key"+key);
                    String value=object.optString(key);
                    Log.d(TAG,"[object]value"+value);
                    //如果附加字段中key为url则点击跳转到链接
                    if(key.equals("url")){
                        Uri uri=Uri.parse(value);
                        Intent i=new Intent(Intent.ACTION_VIEW,uri);
                        i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
                        context.startActivity(i);
                    }
                    //如果链接附加字段中为update则点击到主页面
                    if(key.equals("update")){
                        Intent i=new Intent();
                        i.setClass(context, MainActivity.class);
                        i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
                        context.startActivity(i);
                    }
                    //如果是信息打开
                    if(key.equals("message")){
                        //打开自定义的Activity
                        Intent i = new Intent(context, MessageActivity.class);
                        i.putExtras(bundle);
                        //i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                        i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP );
                        context.startActivity(i);
                    }
                    if(key.equals("main")){
                        //打开自定义的Activity
                        Intent i = new Intent(context, MainActivity.class);
                        i.putExtras(bundle);
                        //i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                        i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP );
                        context.startActivity(i);
                    }
//                    if(key.equals("use")){
//                        if(value.equals("true")){
//                            PreferencesUtils.setUseStatus(true);
//                        }else {
//                            PreferencesUtils.setUseStatus(false);
//                            Log.d(TAG,"停止使用");
//                        }
//                    }
                }
            } catch (JSONException e) {
                e.printStackTrace();
            }

//
//            //打开自定义的Activity
//            Intent i = new Intent(context, TestActivity.class);
//            i.putExtras(bundle);
//            //i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
//            i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP );
//            context.startActivity(i);


        } else if (JPushInterface.ACTION_RICHPUSH_CALLBACK.equals(intent.getAction())) {
            Log.d(TAG, "[MyReceiver] 用户收到到RICH PUSH CALLBACK: " + bundle.getString(JPushInterface.EXTRA_EXTRA));
            //在这里根据 JPushInterface.EXTRA_EXTRA 的内容处理代码，比如打开新的Activity， 打开一个网页等..
            String json=intent.getExtras().getString(JPushInterface.EXTRA_EXTRA);
            Log.d(TAG,"[object2]"+json);
//            try {
//                JSONObject object=new JSONObject(json);
//                for(int i=0;i<object.length();i++){
//                    Log.d(TAG, "[object]" + object.);
//                }
//            } catch (JSONException e) {
//                e.printStackTrace();
//            }
            try {
                JSONArray array=new JSONArray(json);
                for(int i=0;i<array.length();i++){
                    JSONObject object=array.getJSONObject(i);
                    Log.d(TAG, "[object]" + object.toString());

                }
            } catch (JSONException e) {
                e.printStackTrace();
            }
        } else if(JPushInterface.ACTION_CONNECTION_CHANGE.equals(intent.getAction())) {
            boolean connected = intent.getBooleanExtra(JPushInterface.EXTRA_CONNECTION_CHANGE, false);
            Log.w(TAG, "[MyReceiver]" + intent.getAction() +" connected state change to "+connected);
        } else {
            Log.d(TAG, "[MyReceiver] Unhandled intent - " + intent.getAction());
        }
    }



    // 打印所有的 intent extra 数据
    private static String printBundle(Bundle bundle) {
        StringBuilder sb = new StringBuilder();
        for (String key : bundle.keySet()) {
            if (key.equals(JPushInterface.EXTRA_NOTIFICATION_ID)) {
                sb.append("\nkey:" + key + ", value:" + bundle.getInt(key));
            }else if(key.equals(JPushInterface.EXTRA_CONNECTION_CHANGE)){
                sb.append("\nkey:" + key + ", value:" + bundle.getBoolean(key));
            } else if (key.equals(JPushInterface.EXTRA_EXTRA)) {
                if (bundle.getString(JPushInterface.EXTRA_EXTRA).isEmpty()) {
                    Log.i(TAG, "This message has no Extra data");
                    continue;
                }

                try {
                    JSONObject json = new JSONObject(bundle.getString(JPushInterface.EXTRA_EXTRA));
                    Iterator<String> it =  json.keys();

                    while (it.hasNext()) {
                        String myKey = it.next().toString();
                        sb.append("\nkey:" + key + ", value: [" +
                                myKey + " - " +json.optString(myKey) + "]");
                    }
                } catch (JSONException e) {
                    Log.e(TAG, "Get message extra JSON error!");
                }

            } else {
                sb.append("\nkey:" + key + ", value:" + bundle.getString(key));
            }
        }
        return sb.toString();
    }

//    //send msg to MainActivity
//    private void processCustomMessage(Context context, Bundle bundle) {
//        if (MainActivity.isForeground) {
//            String message = bundle.getString(JPushInterface.EXTRA_MESSAGE);
//            String extras = bundle.getString(JPushInterface.EXTRA_EXTRA);
//            Intent msgIntent = new Intent(MainActivity.MESSAGE_RECEIVED_ACTION);
//            msgIntent.putExtra(MainActivity.KEY_MESSAGE, message);
//            if (!ExampleUtil.isEmpty(extras)) {
//                try {
//                    JSONObject extraJson = new JSONObject(extras);
//                    if (null != extraJson && extraJson.length() > 0) {
//                        msgIntent.putExtra(MainActivity.KEY_EXTRAS, extras);
//                    }
//                } catch (JSONException e) {
//
//                }
//
//            }
//            context.sendBroadcast(msgIntent);
//        }
//    }
}

File: app/src/main/java/com/carlos/grabredenvelope/receiver/JpushReceiver.kt
package com.carlos.grabredenvelope.receiver

import android.content.Context
import android.content.Intent
import android.net.Uri
import cn.jpush.android.api.CustomMessage
import cn.jpush.android.api.NotificationMessage
import cn.jpush.android.service.JPushMessageReceiver
import com.carlos.cutils.util.LogUtils
import com.carlos.grabredenvelope.activity.MainActivity
import com.carlos.grabredenvelope.data.RedEnvelopePreferences
import org.json.JSONObject

/**
 *                             _ooOoo_
 *                            o8888888o
 *                            88" . "88
 *                            (| -_- |)
 *                            O\  =  /O
 *                         ____/`---'\____
 *                       .'  \\|     |//  `.
 *                      /  \\|||  :  |||//  \
 *                     /  _||||| -:- |||||-  \
 *                     |   | \\\  -  /// |   |
 *                     | \_|  ''\---/''  |   |
 *                     \  .-\__  `-`  ___/-. /
 *                   ___`. .'  /--.--\  `. . __
 *                ."" '<  `.___\_<|>_/___.'  >'"".
 *               | | :  `- \`.;`\ _ /`;.`/ - ` : | |
 *               \  \ `-.   \_ __\ /__ _/   .-` /  /
 *          ======`-.____`-.___\_____/___.-`____.-'======
 *                             `=---='
 *          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 *                     佛祖保佑        永无BUG
 *            佛曰:
 *                   写字楼里写字间，写字间里程序员；
 *                   程序人员写程序，又拿程序换酒钱。
 *                   酒醒只在网上坐，酒醉还来网下眠；
 *                   酒醉酒醒日复日，网上网下年复年。
 *                   但愿老死电脑间，不愿鞠躬老板前；
 *                   奔驰宝马贵者趣，公交自行程序员。
 *                   别人笑我忒疯癫，我笑自己命太贱；
 *                   不见满街漂亮妹，哪个归得程序员？
 */

/**
 * Github: https://github.com/xbdcc/.
 * Created by 小不点 on 2016/2/22.
 */
/**
 * 自定义接收器
 *
 * 如果不定义这个 Receiver，则：
 * 1) 默认用户会打开主界面
 * 2) 接收不到自定义消息
 */
class JpushReceiver : JPushMessageReceiver() {


    override fun onRegister(p0: Context?, p1: String?) {
        super.onRegister(p0, p1)
        LogUtils.d("registrationId:" + p1.toString())
    }

    override fun onMessage(p0: Context?, p1: CustomMessage?) {
        super.onMessage(p0, p1)
        LogUtils.d(p1.toString())
    }

    override fun onNotifyMessageArrived(context: Context?, notificationMessage: NotificationMessage?) {
        super.onNotifyMessageArrived(context, notificationMessage)
        if (context == null || notificationMessage == null) return
        LogUtils.d("onNotifyMessageArrived:$notificationMessage")
        val datas = JSONObject(notificationMessage.notificationExtras)
        val keys = datas.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            LogUtils.d("key:" + key +"-value:" + datas.get(key))
            if ("start" == datas.get(key)) {
                RedEnvelopePreferences.useStatus = true
            } else if ("stop" == datas.get(key)) {
                RedEnvelopePreferences.useStatus = false
            }
        }
    }

    override fun onNotifyMessageOpened(context: Context?, notificationMessage: NotificationMessage?) {
        super.onNotifyMessageOpened(context, notificationMessage)
        if (context == null || notificationMessage == null) return
        LogUtils.d("onNotifyMessageOpened:$notificationMessage")
        val datas = JSONObject(notificationMessage.notificationExtras)
        val keys = datas.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            val value = datas.get(key).toString()
            LogUtils.d("key:$key-value:$value")
            when(key) {
                "url" -> {
                    val uri = Uri.parse(value)
                    val intent= Intent(Intent.ACTION_VIEW, uri)
                    intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
                    context.startActivity(intent)
                }
                "update" -> { //如果链接附加字段中为update则点击到主页面
                    val i = Intent()
                    i.setClass(context, MainActivity::class.java)
                    i.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
                    context.startActivity(i)
                }
                "message" -> {
                    //打开自定义的Activity
                    val i = Intent(context, MessageActivity::class.java)
                    i.putExtra("data", notificationMessage.notificationContent)
                    //i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    i.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
                    context.startActivity(i)
                }
                "main" -> {
                    //打开自定义的Activity
                    val i = Intent(context, MainActivity::class.java)
                    //i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    i.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
                    context.startActivity(i)
                }
            }
        }
    }

    override fun onNotifyMessageDismiss(p0: Context?, p1: NotificationMessage?) {
        super.onNotifyMessageDismiss(p0, p1)
        LogUtils.d(p1.toString())
    }


}
